crates/ra_ide/src/extend_selection.rs

```diff
@@ -88,6 +104,72 @@ fn try_extend_selection(root: &SyntaxNode, range: TextRange) -> Option<TextRange
     node.parent().map(|it| it.text_range())
 }
 
+fn extend_tokens_from_range(
+    db: &RootDatabase,
+    file_id: FileId,
+    macro_call: ast::MacroCall,
+    original_range: TextRange,
+) -> Option<TextRange> {
+    // compute original mapped token range
+    let mut expanded = None;
+    let range = macro_call
+        .syntax()
+        .descendants_with_tokens()
+        .filter_map(|n| match n {
+            NodeOrToken::Token(token) if token.text_range().is_subrange(&original_range) => {
+                let node = descend_into_macros(db, file_id, token);
```

Comment:
Yeah, I think the problem is that we try to descend into macros for every token in a macro call.

I would imagine we should do this instead:

* find the *first* and the *last* tokens of the selection
* map them forwards
* run extend selection on the new range
* similary, take first and last tokens of the result and map them back to the original invocation.

Looks like a useful abstradction here might be `InFile<(SyntaxToken, SyntaxToken)>` -- a token range we are trying to expand.